src/OVAL/oval_session.c

```diff
@@ -0,0 +1,463 @@
+
+	struct {
+		const char *results;
+		const char *report;
+	} export;
+
+	struct {
+		/* it's called when there is something invalid in input/output files */
+		xml_reporter xml_fn;
+	} reporter;
+
+	bool validation;
+	bool full_validation;
+};
+
+struct oval_session *oval_session_new(const char *filename)
+{
+	oscap_document_type_t scap_type;
+	struct oval_session *session;
+
+	session = (struct oval_session *) oscap_calloc(1, sizeof(struct oval_session));
+
+	session->source = oscap_source_new_from_file(filename);
+	if ((scap_type = oscap_source_get_scap_type(session->source)) == OSCAP_DOCUMENT_UNKNOWN) {
+		oval_session_free(session);
+		return NULL;
+	}
+
+	if (scap_type != OSCAP_DOCUMENT_OVAL_DEFINITIONS && scap_type != OSCAP_DOCUMENT_SDS) {
+		oscap_seterr(OSCAP_EFAMILY_OSCAP, "Session input file was determined but it"
+				" isn't an OVAL file nor a source datastream file.");
+		oval_session_free(session);
+		return NULL;
+	}
+
+	return session;
+}
+
+void oval_session_set_variables(struct oval_session *session, const char *filename)
+{
+	__attribute__nonnull__(session);
+
+	oscap_source_free(session->oval.variables);
+
+	if (filename != NULL)
+		session->oval.variables = oscap_source_new_from_file(filename);
+	else
+		session->oval.variables = NULL;	/* reset */
+}
+
+void oval_session_set_directives(struct oval_session *session, const char *filename)
+{
+	__attribute__nonnull__(session);
+
+	oscap_source_free(session->oval.directives);
+
+	if (filename != NULL)
+		session->oval.directives = oscap_source_new_from_file(filename);
+	else
+		session->oval.directives = NULL;
+}
+
+void oval_session_set_validation(struct oval_session *session, bool validate, bool full_validation)
+{
+	__attribute__nonnull__(session);
+
+	session->validation = validate;
+	session->full_validation = full_validation;
+}
+
+void oval_session_set_datastream_id(struct oval_session *session, const char *id)
+{
+	__attribute__nonnull__(session);
+
+	oscap_free(session->datastream_id);
+	session->datastream_id = oscap_strdup(id);
+}
+
+void oval_session_set_component_id(struct oval_session *session, const char *id)
+{
+	__attribute__nonnull__(session);
+
+	oscap_free(session->component_id);
+	session->component_id = oscap_strdup(id);
+}
+
+void oval_session_set_results_export(struct oval_session *session, const char *filename)
+{
+	__attribute__nonnull__(session);
+
+	oscap_free(session->export.results);
+	session->export.results = oscap_strdup(filename);
+}
+
+void oval_session_set_report_export(struct oval_session *session, const char *filename)
+{
+	__attribute__nonnull__(session);
+
+	oscap_free(session->export.report);
+	session->export.report = oscap_strdup(filename);
+}
+
+void oval_session_set_xml_reporter(struct oval_session *session, xml_reporter fn)
+{
+	__attribute__nonnull__(session);
+
+	session->reporter.xml_fn = fn;
+}
+
+static bool oval_session_validate(struct oval_session *session, struct oscap_source *source, oscap_document_type_t type)
```

Comment:
I think you should be able to drop `type` from `oval_session_validate`. But this can be done any time later. I think we can merge it like it is.